File: tests/client_compatibility/go/connection_test.go
package main

import (
	"context"
	"fmt"
	"os"
	"testing"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

/**
 * Test basic pgx connection to IRIS PGWire server.
 *
 * Tests P0 Handshake Protocol:
 * - SSL negotiation
 * - StartupMessage processing
 * - Authentication
 * - ReadyForQuery state
 */

func getConnectionConfig() string {
	host := os.Getenv("PGWIRE_HOST")
	if host == "" {
		host = "localhost"
	}

	port := os.Getenv("PGWIRE_PORT")
	if port == "" {
		port = "5432"
	}

	database := os.Getenv("PGWIRE_DATABASE")
	if database == "" {
		database = "USER"
	}

	user := os.Getenv("PGWIRE_USERNAME")
	if user == "" {
		user = "test_user"
	}

	password := os.Getenv("PGWIRE_PASSWORD")
	if password == "" {
		password = "test"
	}

	return fmt.Sprintf("postgres://%s:%s@%s:%s/%s?sslmode=disable",
		user, password, host, port, database)
}

func TestBasicConnection(t *testing.T) {
	// GIVEN: PGWire server is running
	ctx := context.Background()
	connString := getConnectionConfig()

	// WHEN: Attempting to connect
	conn, err := pgx.Connect(ctx, connString)
	require.NoError(t, err, "should establish connection")
	defer conn.Close(ctx)

	// THEN: Connection should be established
	require.NotNil(t, conn)

	// Verify with simple query
	var result int
	err = conn.QueryRow(ctx, "SELECT 1").Scan(&result)
	require.NoError(t, err)
	assert.Equal(t, 1, result)
}

func TestConnectionString(t *testing.T) {
	// GIVEN: Connection string
	ctx := context.Background()
	connString := getConnectionConfig()

	// WHEN: Connecting with connection string
	conn, err := pgx.Connect(ctx, connString)
	require.NoError(t, err, "should connect with connection string")
	defer conn.Close(ctx)

	// THEN: Connection succeeds
	var result int
	err = conn.QueryRow(ctx, "SELECT 1").Scan(&result)
	require.NoError(t, err)
	assert.Equal(t, 1, result)
}

func TestConnectionPooling(t *testing.T) {
	// GIVEN: Connection pool configured
	ctx := context.Background()
	connString := getConnectionConfig()

	config, err := pgxpool.ParseConfig(connString)
	require.NoError(t, err)

	config.MaxConns = 10
	config.MinConns = 1

	pool, err := pgxpool.NewWithConfig(ctx, config)
	require.NoError(t, err, "should create connection pool")
	defer pool.Close()

	// WHEN: Acquiring connections from pool
	conn1, err := pool.Acquire(ctx)
	require.NoError(t, err)
	defer conn1.Release()

	conn2, err := pool.Acquire(ctx)
	require.NoError(t, err)
	defer conn2.Release()

	// THEN: Both connections should work
	var result1, result2 int
	err = conn1.QueryRow(ctx, "SELECT 1").Scan(&result1)
	require.NoError(t, err)
	assert.Equal(t, 1, result1)

	err = conn2.QueryRow(ctx, "SELECT 2").Scan(&result2)
	require.NoError(t, err)
	assert.Equal(t, 2, result2)
}

func TestMultipleSequentialConnections(t *testing.T) {
	// GIVEN: Connection configuration
	ctx := context.Background()
	connString := getConnectionConfig()

	// WHEN: Opening and closing connections sequentially
	for i := 0; i < 3; i++ {
		conn, err := pgx.Connect(ctx, connString)
		require.NoError(t, err, "connection %d should succeed", i)

		// THEN: Each connection should succeed
		var result int
		err = conn.QueryRow(ctx, "SELECT $1::int", i+1).Scan(&result)
		require.NoError(t, err)
		assert.Equal(t, i+1, result)

		conn.Close(ctx)
	}
}

func TestServerInformation(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	connString := getConnectionConfig()

	conn, err := pgx.Connect(ctx, connString)
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Querying server version
	var version string
	err = conn.QueryRow(ctx, "SELECT version()").Scan(&version)
	require.NoError(t, err)

	// THEN: Version should be available
	assert.NotEmpty(t, version)
	t.Logf("Server version: %s", version)
}

func TestConnectionErrorHandling(t *testing.T) {
	// GIVEN: Invalid connection configuration
	ctx := context.Background()
	invalidConnString := "postgres://test_user:test@localhost:9999/USER?sslmode=disable"

	// WHEN: Attempting to connect
	conn, err := pgx.Connect(ctx, invalidConnString)

	// THEN: Should return connection error
	assert.Error(t, err, "should fail to connect to invalid port")
	if conn != nil {
		conn.Close(ctx)
	}
}

func TestConnectionTimeout(t *testing.T) {
	// GIVEN: Connection with timeout
	ctx := context.Background()
	connString := getConnectionConfig()

	config, err := pgxpool.ParseConfig(connString)
	require.NoError(t, err)

	config.MaxConnLifetime = 30 * time.Second
	config.MaxConnIdleTime = 10 * time.Second

	pool, err := pgxpool.NewWithConfig(ctx, config)
	require.NoError(t, err)
	defer pool.Close()

	// WHEN: Using connection with timeout
	conn, err := pool.Acquire(ctx)
	require.NoError(t, err)
	defer conn.Release()

	// THEN: Connection should work with timeout settings
	var result int
	err = conn.QueryRow(ctx, "SELECT 1").Scan(&result)
	require.NoError(t, err)
	assert.Equal(t, 1, result)
}

File: tests/client_compatibility/nodejs/connection.test.js
import pg from 'pg';
const { Client, Pool } = pg;

/**
 * Test basic node-postgres connection to IRIS PGWire server.
 *
 * Tests P0 Handshake Protocol:
 * - SSL negotiation
 * - StartupMessage processing
 * - Authentication
 * - ReadyForQuery state
 */

const getConnectionConfig = () => ({
  host: process.env.PGWIRE_HOST || 'localhost',
  port: parseInt(process.env.PGWIRE_PORT || '5432'),
  database: process.env.PGWIRE_DATABASE || 'USER',
  user: process.env.PGWIRE_USERNAME || 'test_user',
  password: process.env.PGWIRE_PASSWORD || 'test',
  ssl: false, // Plain text for testing
});

describe('Basic Connection Tests', () => {
  test('should establish basic connection', async () => {
    // GIVEN: PGWire server is running
    const client = new Client(getConnectionConfig());

    // WHEN: Attempting to connect
    await client.connect();

    try {
      // THEN: Connection should be established
      expect(client).toBeDefined();

      // Verify with simple query
      const result = await client.query('SELECT 1');
      expect(result.rows).toHaveLength(1);
      expect(result.rows[0]['?column?']).toBe(1);
    } finally {
      await client.end();
    }
  });

  test('should connect with connection string', async () => {
    // GIVEN: Connection string
    const config = getConnectionConfig();
    const connectionString = `postgres://${config.user}:${config.password}@${config.host}:${config.port}/${config.database}`;

    const client = new Client({ connectionString, ssl: false });

    // WHEN: Connecting with connection string
    await client.connect();

    try {
      // THEN: Connection succeeds
      const result = await client.query('SELECT 1');
      expect(result.rows).toHaveLength(1);
    } finally {
      await client.end();
    }
  });

  test('should support connection pooling', async () => {
    // GIVEN: Connection pool configured
    const pool = new Pool({
      ...getConnectionConfig(),
      max: 10,
      min: 1,
    });

    try {
      // WHEN: Acquiring connections from pool
      const client1 = await pool.connect();
      const client2 = await pool.connect();

      // THEN: Both connections should work
      const result1 = await client1.query('SELECT 1');
      const result2 = await client2.query('SELECT 2');

      expect(result1.rows[0]['?column?']).toBe(1);
      expect(result2.rows[0]['?column?']).toBe(2);

      client1.release();
      client2.release();
    } finally {
      await pool.end();
    }
  });

  test('should handle multiple sequential connections', async () => {
    // GIVEN: Connection configuration
    const config = getConnectionConfig();

    // WHEN: Opening and closing connections sequentially
    for (let i = 0; i < 3; i++) {
      const client = new Client(config);
      await client.connect();

      // THEN: Each connection should succeed
      const result = await client.query('SELECT $1::int', [i + 1]);
      console.log(`Connection ${i} columns:`, result.fields.map(f => f.name));
      console.log(`Connection ${i} row:`, JSON.stringify(result.rows[0], null, 2));
      expect(result.rows[0].int4).toBe(i + 1);

      await client.end();
    }
  });

  test('should expose server information', async () => {
    // GIVEN: Active connection
    const client = new Client(getConnectionConfig());
    await client.connect();

    try {
      // WHEN: Querying server version
      const result = await client.query('SELECT version()');

      // THEN: Version should be available
      expect(result.rows).toHaveLength(1);
      const version = result.rows[0].version;
      expect(version).toBeDefined();
      console.log('Server version:', version);
    } finally {
      await client.end();
    }
  });

  test('should handle connection errors gracefully', async () => {
    // GIVEN: Invalid connection configuration
    const client = new Client({
      ...getConnectionConfig(),
      port: 9999, // Invalid port
    });

    // WHEN: Attempting to connect
    // THEN: Should throw connection error
    await expect(client.connect()).rejects.toThrow();
  });
});
